File: interactors/src/main/java/fr/coppernic/lib/interactors/ocr/OcrInteractorBuilder.java
package fr.coppernic.lib.interactors.ocr;

import android.content.Context;

public class OcrInteractorBuilder implements OcrInteractor.Builder, OcrInteractor.ContextBuilder {

    private Context context;

    public static OcrInteractor.ContextBuilder get() {
        return new OcrInteractorBuilder();
    }

    @Override
    public OcrInteractor build() {
        return new OcrInteractorIdPlatform(context);
    }

    @Override
    public OcrInteractor.Builder withContext(Context context) {
        this.context = context;
        return this;
    }
}

File: ocr/src/main/java/fr/coppernic/lib/interactors/ocr/OcrInteractorBuilder.kt
package fr.coppernic.lib.interactors.ocr

import android.content.Context
import android.os.Build
import fr.coppernic.sdk.utils.helpers.OsHelper

object OcrInteractorBuilder {
    fun build(context: Context): OcrInteractor {
        return when {
            OsHelper.isIdPlatform() -> {
                ElyctisInteractor(context)
            }
            OsHelper.isCone() -> {
                AccessIsInteractor(context)
            }
            else -> {
                throw OcrInteractorException("Unknown device ${Build.DEVICE} and model ${Build.MODEL}, cannot find suitable OcrInteractor " +
                        "implementation")
            }
        }
    }
}
